MTM-40015: MQTT return codes documentation (#350)
SoftwareAG/c8y-docs: content/device-sdk/mqtt-bundle/implementation.md

> **Info:** The execution of the "last will" updates the device availability.

### MQTT return codes

> When there is an MQTT error, platform responds with a CONNACK message with non-zero return code.

Reviewer: ```suggestion
When there is an MQTT error, the platform responds with a `CONNACK` message with a non-zero return code.
```

This message is the first clue that there is a problem.
Such a return code can be treated similarly to REST API HTTP codes, such as 401.
They can be returned because of an unexpected error, lack of permissions, and so on.

> `CONNACK` is not only a response to `CONNECT` message, but also a way to signal errors by platform.

Reviewer: ```suggestion
`CONNACK` is not only a response to a `CONNECT` message, but also a way to signal errors that occurred in the platform.
```

Therefore, it is possible to receive this message a second time during a normal connection, and without a direct action.

> It is a way to signal a closing connection, as most MQTT clients treat `CONNACK` with a code other than `0` like the connection needs to be closed.

Reviewer: ```suggestion
It is also a way to signal a closing connection, as most MQTT clients treat `CONNACK` with a code other than `0` like the connection needs to be closed.
```

See the details below.

The table below shows the list of errors returned by {{< product-c8y-iot >}}:

|Code|Canonical message|Troubleshooting|
|:-------|:--------|:--------|
|0|Connection accepted | No issue, connection is working.|

> |1|Connection refused, unacceptable protocol version | Unsupported version of MQTT protocol used. Currently, {{< product-c8y-iot >}} only allows 3.1 and 3.1.1.|

Reviewer: ```suggestion
|1|Connection refused, unacceptable protocol version | Unsupported version of the MQTT protocol. Currently, {{< product-c8y-iot >}} only allows 3.1 and 3.1.1.|
```

> |2|Connection refused, identifier rejected | ClientId is not accepted by the platform.|

Reviewer 2: minor: code 2 is returned when the identifier is empty or null, but actually when the client use mqtt in version 3.1 and the client id is longer than 23 characters we still use code 5 due to some generic error handling. I am not sure if we want to add this info to doc or if we need some task to change it, but just wanted to mention it. FYI @karolkaczmarski-sag

Author: I missed that, I think we should include it here

Reviewer 3: Good catch @micst96, @pawelkowalczyksag please add this

> |3|Connection refused, Server unavailable | General platform side error, used on internal errors and unknown authorization problems. <br>Can be received on network issues. <br>Error should be temporary and independent of device state, therefore the usual solution to this is to try again later.|

Reviewer: ```suggestion
|3|Connection refused, Server unavailable | General platform side error, used on internal errors and unknown authorization problems. <br>Can be received on network issues. <br>The error should be temporary and independent of device state, therefore the usual solution to this is to try again later.|
```

|4|Connection refused, bad username or password | Incorrect credentials (wrong username and/or password, but not on empty password). This error is never returned when authenticating with certificates.|

> |5|Connection refused, not authorized | Device side related problem, used when the device doesn't have permissions or is doing something forbidden. For example, if the client sends malformed messages or tries to do any operation without authenticating first, such as publishing a message.<br>Thrown on any issue with certificate authentication (for example, wrong common name, failed auto registration). <br>Also thrown on general issues with receiving device data or some other authorization problem related to the device state on the platform. For example, device Managed Object problems, or the sudden removal of permissions. In this situation it may be required to take action on the platform to investigate and apply a fix.<br> Lastly, it can also be thrown on unexpected exceptions like performance issues, especially during connection. Therefore it is a good approach to repeat connection few times to overcome temporary performance issues.|

Reviewer: ```suggestion
|5|Connection refused, not authorized | Device side related problem, used when the device doesn't have permissions or is doing something forbidden. For example, if the client sends malformed messages or tries to execute an operation without authenticating first, such as publishing a message.<br>Thrown on any issue with certificate authentication (for example, wrong common name, failed auto registration). <br>Also thrown on general issues with receiving device data or some other authorization problem related to the device state on the platform. For example, device managed object problems, or the sudden removal of permissions. In this situation it may be required to take action on the platform to investigate and apply a fix.<br> Lastly, it can also be thrown on unexpected exceptions like performance issues, especially during connection. Therefore it is a good approach to repeat the connection a few times to overcome temporary performance issues.|
```

> |5|Connection refused, not authorized | Device side related problem, used when the device doesn't have permissions or is doing something forbidden. For example, if the client sends malformed messages or tries to execute an operation without authenticating first, such as publishing a message.<br>Thrown on any issue with certificate authentication (for example, wrong common name, failed auto registration). <br>Also thrown on general issues with receiving device data or some other authorization problem related to the device state on the platform. For example, device managed object problems, or the sudden removal of permissions. In this situation it may be required to take action on the platform to investigate and apply a fix.<br> Lastly, it can also be thrown on unexpected exceptions like performance issues, especially during connection. Therefore it is a good approach to repeat the connection a few times to overcome temporary performance issues.|

Reviewer 3: Not only "device side" e.g. failures (for whatever reason) during processing the message.

> Refer to the [MQTT Version 3.1.1 > 3.2 CONNACK - Acknowledge connection request](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718033) for details on the official MQTT connection return codes.

Reviewer: ```suggestion
Refer to [MQTT Version 3.1.1 > 3.2 CONNACK - Acknowledge connection request](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718033) for details on the official MQTT connection return codes.
```

> **Info:** The execution of the "last will" updates the device availability.

### MQTT return codes

> When initiating an MQTT connection, you can be exposed to several errors.

Reviewer 4: an -> a ?

Author: @MWindel ? I actually don't the rules, but I never saw "A MQTT" always "AN MQTT"

Reviewer: it's "an"

These are presented by the platform with a `CONNACK` message via a return code, before closing connection.
This message is the first clue that there is a problem.
Such a return code can be treated similarly to REST API HTTP codes, such as 401.
They can be returned because of an unexpected error, lack of permissions, and so on.

> The platform can also use `CONNACK` to signal a sudden error, such as a network issue.

Author: Remove sudden error, it is just a way of signaling errors by platform

Therefore, it is possible to receive this message a second time during a normal connection, and without a direct action.

> It is a way to signal a closing connection, as most clients treat `CONNACK` with a code other than `0` like the connection needs to be closed.

Reviewer 4: _as most *mqtt* clients_
or 
_as most *mqtt client applications*_

See the details below.

The table below shows the list of errors returned by {{< product-c8y-iot >}}:

|Code|Canonical message|Troubleshooting|
|:-------|:--------|:--------|
|0|Connection accepted | No issue, connection is working.|
|1|Connection refused, unacceptable protocol version | Unsupported version of MQTT protocol used. Currently, {{< product-c8y-iot >}} only allows 3.1 and 3.1.1.|

> |2|Connection refused, identifier rejected | ClientId is not accepted by the platform.|

Author: If ClientId is empty, You won't get this error I think

> |2|Connection refused, identifier rejected | ClientId is not accepted by the platform.|

Reviewer 3: Maybe clientid will be rejected earlier, so this "return" would not execute - please check :)

> |3|Connection refused, Server unavailable | General platform side error, used on internal errors and unknown authorization problems. <br>Can be received on network issues. <br>Error should be temporary and independent of device state, therefore the solution to this is to try again later. <br>See special cases below.|

Reviewer 3: Please write description for codes 3,4,5 as if tenant options are always false (we don't want to document such workarounds)
Compare with https://docs.google.com/document/d/1QZS7vEk_r0ta2TskfwpjclEy8qW0xvjF6EPjLEJ6LzA/edit#

> |4|Connection refused, bad user name or password | Incorrect credentials (wrong username and/or password, but not on empty password. See special cases below). This error is never returned when authenticating with certificates.|

Reviewer 4: why we have in same line `user name` and `username`? Can it be standarized?

> |5|Connection refused, not authorized | Device side related problem, used when the device doesn't have permissions or is doing something forbidden. For example, if the client sends malformed messages or tries to do any operation without authenticating first, such as publishing a message.<br>Thrown on any issue with certificate authentication (for example, wrong common name, failed auto registration). <br>Also thrown on general issues with receiving device data or some other authorization problem related to the device state on the platform. For example, device managed object problems, or the sudden removal of permissions. In this situation it may be required to take action on the platform to investigate and apply a fix. By default, it can also be thrown after connecting, which signals unexpected exceptions like performance issues.<br>See special cases below.|

Reviewer 4: `For example, device managed object problems, or the sudden removal of permissions.`

What do we mean by `device managed object problems`? Malformed json, missing externalID or agent fragment for operations or something else?

Author: We talked about this on chat, its not really clear what can cause a problem. But this more about giving user a clue - go check managed object, maybe there's a problem

> Due to inconsistencies in the {{< product-c8y-iot >}} implementation of MQTT return codes, there are special cases:

Reviewer 2: I would not point to inconsistencies in c8y implementation in public doc (that to some extent is also true), becasue the main problem was that MQTT in 3.1.x version doesn't support all necessary codes - it is solved in MQTT 5.0 version. @karolkaczmarski-sag what do you think?
